x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java

```diff
@@ -64,427 +113,1139 @@
+                    regexBuilder.append("\\b");
+                }
+                prevLetterGroup = letterGroup;
+            } else {
+                if (PUNCTUATION_THAT_NEEDS_ESCAPING_IN_REGEX.indexOf(curChar) >= 0) {
+                    grokPatternBuilder.append('\\');
+                    regexBuilder.append('\\');
+                }
+                grokPatternBuilder.append(curChar);
+                regexBuilder.append(curChar);
+            }
+
+            prevChar = curChar;
+            ++pos;
+        }
+
+        if (prevLetterGroup == null) {
+            throw new IllegalArgumentException("No time format letter groups in override format [" + overrideFormat + "]");
+        }
+
+        return new Tuple<>(grokPatternBuilder.toString(), regexBuilder.toString());
     }
 
     /**
-     * Find the first timestamp format that matches part of the supplied value,
-     * excluding a specified number of candidate formats.
-     * @param text The value that the returned timestamp format must exist within.
-     * @param ignoreCandidates The number of candidate formats to exclude from the search.
+     * Given a user supplied Java timestamp format, return an appropriate candidate timestamp object as required by this class.
+     * The returned candidate might be a built in one, or might be generated from the supplied format.
+     * @param overrideFormat A user supplied Java timestamp format.
      * @param timeoutChecker Will abort the operation if its timeout is exceeded.
-     * @return The timestamp format, or <code>null</code> if none matches.
+     * @return An appropriate candidate timestamp object.
      */
-    public static TimestampMatch findFirstMatch(String text, int ignoreCandidates, TimeoutChecker timeoutChecker) {
-        return findFirstMatch(text, ignoreCandidates, null, timeoutChecker);
+    static CandidateTimestampFormat makeCandidateFromOverrideFormat(String overrideFormat, TimeoutChecker timeoutChecker) {
+
+        // First check for a special format string
+        switch (overrideFormat.toUpperCase(Locale.ROOT)) {
+            case "ISO8601":
+                return ISO8601_CANDIDATE_FORMAT;
+            case "UNIX_MS":
+                return UNIX_MS_CANDIDATE_FORMAT;
+            case "UNIX":
+                return UNIX_CANDIDATE_FORMAT;
+            case "TAI64N":
+                return TAI64N_CANDIDATE_FORMAT;
+        }
+
+        // Next check for a built-in candidate that incorporates the override, and prefer this
+
+        // If the override is not a valid format then one or other of these two calls will
+        // throw, and that is how we'll report the invalid format to the user
+        Tuple<String, String> grokPatternAndRegex = overrideFormatToGrokAndRegex(overrideFormat);
+        DateTimeFormatter javaTimeFormatter = DateTimeFormatter.ofPattern(overrideFormat, Locale.ROOT);
+
+        // This timestamp is chosen to avoid the possibility of truncating zeroes
```

Comment:
I'm not quite getting why 29th Feb is doing the trick. Could you elaborate?

Reply:
Well spotted.  This is a bug.  The month, day and hour should all be less than 10 to distinguish formats that don't use leading zeroes for them.  I'll fix the comment too.